src/superqt/utils/_throttler.py

```diff
@@ -0,0 +1,364 @@
+    timeoutChanged = Signal(int)
+    timerTypeChanged = Signal(Qt.TimerType)
+
+    def __init__(
+        self,
+        kind: Kind,
+        emissionPolicy: EmissionPolicy,
+        parent: Optional[QObject] = None,
+    ) -> None:
+        super().__init__(parent)
+
+        self._kind = kind
+        self._emissionPolicy = emissionPolicy
+        self._hasPendingEmission = False
+
+        self._timer = QTimer()
+        self._timer.setSingleShot(True)
+        self._timer.setTimerType(Qt.TimerType.PreciseTimer)
+        self._timer.timeout.connect(self._maybeEmitTriggered)
+
+    def kind(self) -> Kind:
+        """Return the kind of throttler (throttler or debouncer)."""
+        return self._kind
+
+    def emissionPolicy(self) -> EmissionPolicy:
+        """Return the emission policy (trailing or leading)."""
+        return self._emissionPolicy
+
+    def timeout(self) -> int:
+        """Return current timeout in milliseconds."""
+        return self._timer.interval()  # type: ignore
+
+    def setTimeout(self, timeout: int) -> None:
+        """Set timeout in milliseconds"""
+        if self._timer.interval() != timeout:
+            self._timer.setInterval(timeout)
+            self.timeoutChanged.emit(timeout)
+
+    def timerType(self) -> Qt.TimerType:
+        """Return current Qt.TimerType."""
+        return self._timer.timerType()
+
+    def setTimerType(self, timerType: Qt.TimerType) -> None:
+        """Set current Qt.TimerType."""
+        if self._timer.timerType() != timerType:
+            self._timer.setTimerType(timerType)
+            self.timerTypeChanged.emit(timerType)
+
+    def throttle(self) -> None:
+        """Emit triggered if not running, then start timer."""
+        # public slot
+        self._hasPendingEmission = True
+        # Emit only if we haven't emitted already. We know if that's
+        # the case by checking if the timer is running.
+        if (
+            self._emissionPolicy is EmissionPolicy.Leading
+            and not self._timer.isActive()
+        ):
+            self._emitTriggered()
+
+        # The timer is started in all cases. If we got a signal, and we're Leading,
+        # and we did emit because of that, then we don't re-emit when the timer fires
+        # (unless we get ANOTHER signal).
+        if self._kind is Kind.Throttler:  # sourcery skip: merge-duplicate-blocks
+            if not self._timer.isActive():
+                self._timer.start()  # actual start, not restart
+        elif self._kind is Kind.Debouncer:
+            self._timer.start()  # restart
+
+        assert self._timer.isActive()
+
+    def cancel(self) -> None:
+        """ "Cancel and pending emissions."""
+        self._hasPendingEmission = False
+
+    def flush(self) -> None:
+        """ "Force emission of any pending emissions."""
+        self._maybeEmitTriggered()
+
+    def _emitTriggered(self) -> None:
+        self._hasPendingEmission = False
+        self.triggered.emit()
```

Comment:
```suggestion
        self.triggered.emit()
        self._timer.start()
```
I'm not sure but maybe this will allow avoiding double lines visible in throttle leading?

Reply:
thanks lemme check

Reply:
indeed!  thanks @Czaki!